metrics/lastseen/tracker.go

```diff
@@ -0,0 +1,226 @@
+package lastseen
+
+import (
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/wrapp/instrumentation/logs"
+)
+
+// Tracker takes care of tracking and exporting metrics
+type Tracker interface {
+	// SetSeen records a seen timestamp for the given field
+	SetSeen(ctx context.Context, field string)
+	// GetSeen returns the unix time stamp when the given field was last tracked as seen
+	GetSeen(ctx context.Context, field string) int64
+	// Wait for all exports to finish
+	WaitForExporters()
+}
+
+// LastSeen keeps track of a last seen and its last flush time
+type LastSeen struct {
+	Value     int64
+	LastFlush int64
+}
+
+// TrackerOption defines a configuration function
+type TrackerOption func(*TrackerOptions)
+
+// TrackerOptions defines options for the tracker
+type TrackerOptions struct {
+	flushIntervalSeconds int64
+	useTicker            bool
+	exporterFactories    []ExporterFactory
+}
+
+type tracker struct {
+	seens                sync.Map
+	exporterMu           sync.Mutex
+	exporter             Exporter
+	wg                   sync.WaitGroup
+	flushIntervalSeconds int64
+	useTicker            bool
+}
+
+var (
+	once            sync.Once
+	lastSeenTracker Tracker
+)
+
+// NewTrackerSingleton initializes a last seen tracker singleton
+func NewTrackerSingleton(ctx context.Context, options ...TrackerOption) (Tracker, error) {
+
+	opts := TrackerOptions{
+		flushIntervalSeconds: 30,
+		useTicker:            false,
+	}
+	for _, apply := range options {
+		apply(&opts)
+	}
+
+	var exporters []Exporter
+	for _, factory := range opts.exporterFactories {
+		exporter, err := factory()
+		if err != nil {
+			return nil, err
+		}
+		exporters = append(exporters, exporter)
+	}
+
+	once.Do(func() {
+
+		t := &tracker{
+			useTicker:            opts.useTicker,
+			flushIntervalSeconds: opts.flushIntervalSeconds,
+			exporter:             Multi(exporters...),
+		}
+
+		if opts.useTicker {
+			t.start()
+		}
+
+		lastSeenTracker = t
+	})
+	return lastSeenTracker, nil
+}
+
+// GetTracker gets the last seen tracker singleton
+func GetTracker() Tracker {
+	if lastSeenTracker == nil {
+		return nil
+	}
```

Comment:
Shouldn't we  throw an error if there is no tracker defined ? The clients would need to deal with the `nil` check anyway so we the `if` will be there no matter what.

Reply:
Added a global SetSeen method, then we don't need to call GetTracker. If no tracker has been configured, nothing is tracked.